### PR TITLE
[test] Fix for kernel names issues in permutation iterator tests

### PR DESCRIPTION
**Pull request by danhoeflinger**
This PR is being proposed as an alternative to #1370

According to the SYCL 2020 spec [Section 5.2 Naming of kernels](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:naming.kernels), "kernel names must be forward declarable at namespace scope".  
It seems that the tags we were using to signify test types for permutation iterator map methods, which are a group of structs inside a struct do not meet this criteria.  

Changing them to be base level structures with the outer name incorporated resolves this issue.

We also have a few kernel naming clashes for `permutation_iterator_parallel_for.pass` and `permutation_iterator_parallel_merge.pass`.  This was fixed with help from @SergeyKopienko and #1370, to add utility functions to wrap of kernel names for hetero policies and just pass through host policies.

### CODE REVIEW

**Review by SergeyKopienko — APPROVED** [review]
LGTM, if it's compiler with unnamed lambda off

**Review by mmichel11 — DISMISSED** [review]
LGTM

**Comment by danhoeflinger** [comment]
> Looks like not all tests compiled now:
> 
> * for example. `test\parallel_api\iterator\permutation_iterator_parallel_for.pass.cpp`

I've added name clash fixed for permutation_iterator_parallel_for and permutation_iterator_parallel_merge.  I believe they all compile and run successfully with unnamed lambda off (though independent confirmation is encouraged).

**Comment by danhoeflinger** [comment]
Fixed an issue pointed out by @SergeyKopienko offline of waiting on incorrect policy.

**Review by SergeyKopienko — DISMISSED** [review]
LGTM

**Review by SergeyKopienko — DISMISSED** [review]
LGTM

**Review by mmichel11 — DISMISSED** [review]
The changes since my initial approval LGTM. Let's hold off on merging until the issues in main are resolved.

**Comment by danhoeflinger** [comment]
I believe I have successfully rebased to be a descendant of the current main.  
`git rebase -i`, and "accepting current" for merge conflicts during rebase. 

Output from git log (with my pretty print format):
```
6e8276736  (HEAD -> dev/dhoeflin/kernel_naming_permutation_iter, origin/dev/dhoeflin/kernel_naming_permutation_iter) Dan Hoeflinger Fri, 26 Jan 2024 11:12:20 -0500 minor improvements
15e0d1c5e  Dan Hoeflinger Fri, 26 Jan 2024 10:40:55 -0500 fixing waiting on wrong policy
20a10cdd6  Dan Hoeflinger Fri, 26 Jan 2024 10:10:10 -0500 fix for name clash in merge + formatting
df07c420f  Dan Hoeflinger Fri, 26 Jan 2024 09:58:26 -0500 fixing issue with duplicate kernel name
b84e620e8  Dan Hoeflinger Fri, 26 Jan 2024 09:17:43 -0500 fix for kernel names being forward declarable at namespace scope
286097b43  (origin/main, origin/HEAD, main) Dan Hoeflinger Fri, 26 Jan 2024 10:04:46 -0500 Fix zip iterators to be device copyable, fix for require_access (#1371)
a860b5d1b  Dan Hoeflinger Thu, 25 Jan 2024 16:27:11 -0500 Bug fixes for permutation_iterator (#1383)
67fa65c05  Matthew Michel Thu, 25 Jan 2024 14:39:14 -0600 Correct preprocessor SYCL target checks for reduce and min/max element (#1342)
0f16e05d4  Anuya Welling Thu, 25 Jan 2024 14:09:36 -0600 Dynamic selection CI fix (#1384)
3c21893b9  Anuya Welling Thu, 25 Jan 2024 08:16:35 -0600 [Dynamic Selection] Addressed comments to sycl_backend (#1376)
2a69fb0f9  Anuya Welling Thu, 25 Jan 2024 08:15:50 -0600 removed sycl sanity (#1378)
a2c938022  Anuya Welling Thu, 25 Jan 2024 08:15:09 -0600 Addressed comments for auto tune policy (#1379)
3b6132872  Ruslan Arutyunyan Thu, 25 Jan 2024 06:11:30 -0700 Support proxy objects on iterator dereference for transform algorithms (#1233)
e96361828  MikeDvorskiy Wed, 24 Jan 2024 10:41:39 +0100 [oneDPL][doc] + limitation for permutation_iterator usage and for 'Uninitialized storage' algorithms
2bc38d2fb  Dan Hoeflinger Wed, 24 Jan 2024 21:36:14 -0500 Histogram bugfixes for FPGA and icpx 2022.3  (#1367)
...
```  

@mmichel11 @SergeyKopienko  please take a quick look to confirm we are looking good here post `origin/main` rebasing

**Review by mmichel11 — APPROVED** [review]
I took a look through the commits and the rebase looks successful and is consistent with the changes on main.

**Pull request merged**
